commit d2fbc48658022f48625064ae192baff52057987d
Author: Alison Schofield <alison.schofield@intel.com>
Date:   Tue Apr 18 20:26:25 2023 -0700

    cxl/memdev: Add support for the Inject Poison mailbox command
    
    CXL devices optionally support the INJECT POISON mailbox command. Add
    memdev driver support for the mailbox command.
    
    Per the CXL Specification (3.0 8.2.9.8.4.2), after receiving a valid
    inject poison request, the device will return poison when the address
    is accessed through the CXL.mem driver. Injecting poison adds the address
    to the device's Poison List and the error source is set to Injected.
    In addition, the device adds a poison creation event to its internal
    Informational Event log, updates the Event Status register, and if
    configured, interrupts the host.
    
    Also, per the CXL Specification, it is not an error to inject poison
    into an address that already has poison present and no error is
    returned from the device.
    
    If the address is not contained in the device's dpa resource, or is
    not 64 byte aligned, return -EINVAL without issuing the mbox command.
    
    Poison injection is intended for debug only and will be exposed to
    userspace through debugfs. Restrict compilation to CONFIG_DEBUG_FS.
    
    Signed-off-by: Alison Schofield <alison.schofield@intel.com>
    Reviewed-by: Jonathan Cameron <Jonathan.Cameron@huawei.com>
    Reviewed-by: Dave Jiang <dave.jiang@intel.com>
    Link: https://lore.kernel.org/r/241c64115e6bd2effed9c7a20b08b3908dd7be8f.1681874357.git.alison.schofield@intel.com
    Tested-by: Jonathan Cameron <Jonathan.Cameron@huawei.com>
    Signed-off-by: Dan Williams <dan.j.williams@intel.com>

diff --git a/drivers/cxl/cxlmem.h b/drivers/cxl/cxlmem.h
index bfb75bf0182e..2a0625e1d3aa 100644
--- a/drivers/cxl/cxlmem.h
+++ b/drivers/cxl/cxlmem.h
@@ -615,6 +615,11 @@ struct cxl_mbox_poison_out {
 #define CXL_POISON_SOURCE_INJECTED	3
 #define CXL_POISON_SOURCE_VENDOR	7
 
+/* Inject & Clear Poison  CXL 3.0 Spec 8.2.9.8.4.2/3 */
+struct cxl_mbox_inject_poison {
+	__le64 address;
+};
+
 /**
  * struct cxl_mem_command - Driver representation of a memory device command
  * @info: Command information as it exists for the UAPI
@@ -689,6 +694,7 @@ int cxl_poison_state_init(struct cxl_dev_state *cxlds);
 int cxl_mem_get_poison(struct cxl_memdev *cxlmd, u64 offset, u64 len,
 		       struct cxl_region *cxlr);
 int cxl_trigger_poison_list(struct cxl_memdev *cxlmd);
+int cxl_inject_poison(struct cxl_memdev *cxlmd, u64 dpa);
 
 #ifdef CONFIG_CXL_SUSPEND
 void cxl_mem_active_inc(void);